addons/project/models/project_task_type.py

```diff
@@ -79,6 +82,44 @@ def copy(self, default=None):
             default['name'] = _("%s (copy)") % (self.name)
         return super().copy(default)
 
+    def unlink(self): # TODO: write test for deleting stages in batch (+ for multiple user in sudo)
+        # 1. All non-personal stages are processed
+        stages_to_unlink = self.filtered(lambda stage: not stage.user_id)
+
+        # 2. Personal stages are processed if the user still has at least one personal stage after unlink
+        raise_personal_stage_error = False
+        remaining_personal_stages = self.env['project.task.type'].search([('user_id', 'in', self.user_id.ids), ('id', 'not in', self.ids)], order='sequence DESC')
+        for user in self.user_id:
+            user_stages_to_unlink = self.filtered(lambda stage: stage.user_id == user)
+            user_remaining_stages = remaining_personal_stages.filtered(lambda stage: stage.user_id == user)
+            if len(user_remaining_stages) > 0:
+                stages_to_unlink |= user_stages_to_unlink
+                self._prepare_personal_stages_deletion(user_stages_to_unlink, user_remaining_stages)
+            else:
+                raise_personal_stage_error = True
+        result = super(ProjectTaskType, stages_to_unlink).unlink()
+        if raise_personal_stage_error:
+            raise ValidationError(_("Each user should at least have one personal stage. Create a new stage to which the tasks can be transferred after the selected one(s) is deleted."))
+        return result
+
+    def _prepare_personal_stages_deletion(self, stages_to_delete, remaining_stages):
+        """
+        _prepare_personal_stages_deletion prepare the deletion of personal stages of a single user.
+        Tasks using that stage will be moved to the first stage with a lower sequence if it exists
+        higher if not. remaining_stages recordset can not be empty.
+        """
+        remaining_stages_dict = [{'id': stage.id, 'seq': stage.sequence} for stage in remaining_stages]
+        stages_to_delete_dict = sorted([{'id': stage.id, 'seq': stage.sequence} for stage in stages_to_delete],
+                                       key=lambda stage: stage['seq'])
+        replacement_stage_id = remaining_stages_dict.pop()['id']
+        next_replacement_stage = remaining_stages_dict.pop() if len(remaining_stages_dict) > 0 else False
+
+        for stage in stages_to_delete_dict:
+            while next_replacement_stage and next_replacement_stage['seq'] < stage['seq']:
+                replacement_stage_id = next_replacement_stage['id']
+                next_replacement_stage = remaining_stages_dict.pop() if len(remaining_stages_dict) > 0 else False
```

Comment:
> Isn't that equivalent?

I don't think this would work for multiple stages to delete. Indeed, with this solution, for each stage to delete, next_replacement_stage will be set to the next one in line, even if it doesn't need to as the check on the sequence is made after the assignment. Also, if no more replacement_stage in line, replacement_stage_id will not be set, resulting in an error.